src/join.sol

```diff
@@ -427,3 +427,113 @@ contract AuthGemJoin is LibNote {
         require(gem.transfer(usr, wad), "AuthGemJoin/failed-transfer");
     }
 }
+
+interface GemLike7 {
+    // matches ERC20 spec
+    function decimals() external view returns (uint);
+    // matches ERC20 spec
+    function transfer(address,uint) external;
+    // matches ERC20 spec
+    function transferFrom(address,address,uint) external;
+    // USDT uses balanceOf() constant
+    function balanceOf(address) external view returns (uint);
+    // USDT uses allowance() constant
+    function allowance(address,address) external view returns (uint);
+    // doesn't match ERC20 spec
+    function upgradedAddress() external view returns (address);
+    // doesn't match ERC20 spec
+    function deprecated() external view returns (bool);
+    // doesn't match ERC20 spec
+    function setImplementation(address,uint) external;
+}
+
+contract GemJoin7 is LibNote {
+    mapping (address => uint) public wards;
+    function rely(address usr) external note auth { wards[usr] = 1; }
+    function deny(address usr) external note auth { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    VatLike  public vat;
+    bytes32  public ilk;
+    GemLike7 public gem;
+    uint     public dec;
+    uint     public live; // Access flag
+
+    mapping (address => uint256) public implementations;
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        gem = GemLike7(gem_);
+        dec = gem.decimals();
+        require(dec < 18, "GemJoin7/decimals-18-or-higher");
+        wards[msg.sender] = 1;
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        setImplementation(address(gem), 1);
+    }
+
+    function cage() external note auth {
+        live = 0;
+    }
+
+    function setImplementation(address implementation, uint256 permitted) public auth note {
+        implementations[implementation] = permitted; // 1 live, 0 distable
+    }
+
+    function mul(uint x, uint y) internal pure returns (uint z) {
+        require(y == 0 || (z = x * y) / y == x, "GemJoin7/overflow");
+    }
+
+    function join(address urn, uint wad) public note {
+        require(live == 1, "GemJoin7/not-live");
+        // mul does overflow check so require(wad < 2 ** 255) not needed
+        uint wad18 = mul(wad, 10 ** (18 - dec));
+        require(int(wad18) >= 0, "GemJoin7/overflow");
+
+        // implementation check
+        // tether uses a `deprecated` boolean; if deprecated is true, calls are forwarded to
+        // an `upgradedAddress` address. so check if deprecated, if so, require upgradedAddress
+        // to be approved by governance
+        if (gem.deprecated()) {
```

Comment:
I'm not sure this is necessary. As long as we set `implementations[0x000...] == 1` we'd only need to do one external load. If they ever do upgrade then this would perform 2 external loads on every call.

Reply:
removed the if block